### ISSUE TITLE
Problem Allegro API 

### ISSUE THREAD

**Issue by ircoo2022** [allegro/allegro-api#6521]
Witam,

mam pytanie mam zarejestrowaną aplikację ale podczas powiązania konta https://allegro.pl/auth/oauth/authorize?response_type=code&client_id=CLIENT_ID&redirect_uri=LINK_DO_PHP

logiuj się do allegro i mam : Nie możemy wyświetlić strony
Numer błędu: 429e000bcd864064

linmki redirect ok

a inna sprawa czy jak mam jedna aplikacje np. domena.pl/api.php to czy mogę dodawać parametry get np. api.php?akcja=dodaj ?

lub dane post aby allegro mi zwracało chodzi o to żeby miec 1 link allegro api a rozne akcje nim robic aby nie trzabylo 10 redirectów

**Comment by Lukasz-Zurek** [comment]
Przyczyną błędu jest redirect_uri, który nie jest zgodny z tym, który podałeś podczas rejestracji aplikacji. Możesz w każdej chwili go edytować.

Jeśli chodzi o dodawanie parametrów do tego adresu - nie udostępniamy takiej możliwości, więc tutaj musisz niestety podać kilka redirect URIs, aby uniknąć podobnych problemów podczas autoryzacji.

**Comment by ircoo2022** [comment]
URI OK jest

Nie możemy wyświetlić strony
Numer błędu: 88d4212f037afa88

cały czas

**Comment by ircoo2022** [comment]
Czy w REDIRECT_URI możemy podać domenę no. mojadomena.pl i dzieki czemy mamy autoryzacje linkow rozszerzajacych 

np. mojadoman.pl/aa.php mojadomena.pl/bb.php itd ?

**Comment by Lukasz-Zurek** [comment]
@ircoo2022 Nie, musisz zawsze wskazać bezpośredni link. Nie mogą to być ścieżki względne, etc. Ograniczenia w adresach przekierowań znajdziesz pod [tym adresem](https://github.com/allegro/allegro-api/issues/2737).

**Comment by ircoo2022** [comment]
Ok   A czy dobrze myślę ew poprawić mnie , chce aby działała moja Aplikacja ala badelinker     Mam np 10 konto allegro  I chce mieć 1 aplikację z 1 uri redirect   Do tej aplikacji chce podpinać się,  powiązać konto ta 1 aplikacja i dodawac aukcje, oferty do każdego konta allegro osobno   Możliwe to ? 
            
          
     
      
       
        Dnia 28 grudnia 2022 14:15 Lukasz-Zurek ***@***.***&gt; napisał(a):
       
    
       
         
 github.com @ircoo2022  Nie, musisz zawsze wskazać bezpośredni link. Nie mogą to być ścieżki względne, etc. Ograniczenia w adresach przekierowań znajdziesz pod  github.com tym adresem .

**Comment by Lukasz-Zurek** [comment]
Tak, możesz zarejestrować jedną aplikację z jednym redirect-uri, a następnie skonfigurować proces autoryzacji, który pozwala na uwierzytelnienie użytkownika, a tym samym - uzyskanie przyzwolenia na wykonywanie przez aplikację żądań w jego imieniu. Zapoznaj się proszę z [naszym poradnikiem](https://developer.allegro.pl/tutorials/uwierzytelnianie-i-autoryzacja-zlq9e75GdIR#wstep).

**Comment by ircoo2022** [comment]
ok dzieki

ten atrybut state mnie zastanawia

https://allegro.pl/auth/oauth/authorize?response_type=code&client_id=a21...6be&redirect_uri=http://exemplary.redirect.uri


czyli moge do linka autoryzacyjnego nadac https://allegro.pl/auth/oauth/authorize?response_type=code&client_id=a21...6be&redirect_uri=http://exemplary.redirect.uri&state=mojazmienna

i ona zostanei zwrocina razem z codem autoryzacyjnym lub z tokenem ?? tak ?

**Comment by Lukasz-Zurek** [comment]
W parametrze "state" przekazujesz dodatkowe dane, które zostaną przekazane z powrotem do aplikacji po autoryzacji - wraz z kodem autoryzacyjnym potrzebnym do uzyskania pary tokenów dostępowych (krok drugi procesu autoryzacji [Authorization Code Flow](https://developer.allegro.pl/tutorials/uwierzytelnianie-i-autoryzacja-zlq9e75GdIR#autoryzacja-uzytkownika)).

Czyli w pierwszym kroku autoryzacji przekazujesz parametr "state" (opcjonalnie) np. "xxx" wraz z wymaganymi danymi, następnie zwracamy Tobie kod autoryzacyjny oraz przekazany przez Ciebie parametr "state".

np. http://exemplary.redirect.uri?code=1jx217wTtda1chXAP04TTnK8p8I4fmT7&state=xxx

**Comment by ircoo2022** [comment]
ok dzieki

zastanawia mnie waznosc tokena bo mimo że robie wg waszego poradnika to token i tak 12h jest wazny i trzeba go znow autoryzowac pobierac kod itd

if (!empty($_GET["code"])) {
	$refreshToken = getRefreshToken($_GET["code"]);
	$nextToken = $access_token = TokenRefresh($refreshToken);
}

fragment mojego i waszego kodu w sumie,
posługuję się wtedy tym $nextToken = $access_token = TokenRefresh($refreshToken); tokenem i działa też 12h...
jak to zdebugować?

**Comment by Lukasz-Zurek** [comment]
Podstawowy token dostępowy (access_token) zawsze jest ważny 12 godzin. Zapoznaj się proszę z procesem [przedłużania ważności tokena](https://developer.allegro.pl/tutorials/uwierzytelnianie-i-autoryzacja-zlq9e75GdIR#przedluzenie-waznosci-tokena). Refresh token pozwala na uzyskanie nowego tokena dostępowego dla danego użytkownika bez konieczności ingerencji z jego strony.

**Comment by ircoo2022** [comment]
Dokładnie i postępuję wg waszej instrukcji wkleiłem wyżej ten sam kod co w naszej dokumentacji :)

function main(){
    if ($_GET["code"]) {
        $refreshToken = getRefreshToken($_GET["code"]);
        $nextToken= TokenRefresh($refreshToken);
        echo "access_token = ", $nextToken;
    } else {
        getAuthorizationCode();
    }
}

**Comment by Lukasz-Zurek** [comment]
> Dokładnie i postępuję wg waszej instrukcji wkleiłem wyżej ten sam kod co w naszej dokumentacji :)
> 
> function main(){ if ($_GET["code"]) { $refreshToken = getRefreshToken($_GET["code"]); $nextToken= TokenRefresh($refreshToken); echo "access_token = ", $nextToken; } else { getAuthorizationCode(); } }
> 
> czyli najpierw grant_type=authorization_code to wyzej zwraca refresh_token i przekazuję do zapytania grant_type=refresh_token i zwraca access_token i ten powinien byc juz dlugo wazny ale jest tak samo po 12h :)
> 

Tak jak wspomniałem powyżej - podstawowy access_token jest ważny 12h. Procesu przedłużenia ważności tokena nie można rozumieć dosłownie. Za każdym razem gdy używasz refresh token, w odpowiedzi otrzymujesz nową parę:

- access token (ważny kolejne 12h ),
- refresh token (ważny kolejne 3 miesiące). 

Dzięki temu nie musisz za każdym razem od początku się autoryzować. Czyli refresh token pozwala na uzyskanie nowego tokena dostępowego dla danego użytkownika bez konieczności ingerencji z jego strony.

> Nie wiem czy można to jakoś zdebugować po waszej stronie ? jakiś trace-id ?

W tym konkretnym przypadku nie będzie takiej potrzeby. Oczywiście jeśli nadal masz wątpliwości co do całego procesu - jestem do dyspozycji :)

**Comment by ircoo2022** [comment]
Dnia 2 stycznia 2023 14:13 Lukasz-Zurek ***@***.***&gt; napisał(a):   Dokładnie i postępuję wg waszej instrukcji wkleiłem wyżej ten sam kod co w naszej dokumentacji :)  function main(){ if ($_GET[&#34;code&#34;]) { $refreshToken = getRefreshToken($_GET[&#34;code&#34;]); $nextToken= TokenRefresh($refreshToken); echo &#34;access_token = &#34;, $nextToken; } else { getAuthorizationCode(); } }  czyli najpierw grant_type=authorization_code to wyzej zwraca refresh_token i przekazuję do zapytania grant_type=refresh_token i zwraca access_token i ten powinien byc juz dlugo wazny ale jest tak samo po 12h :)  Tak jak wspomniałem powyżej - podstawowy access_token jest ważny 12h. Procesu przedłużenia ważności tokena nie można rozumieć dosłownie. Za każdym razem gdy używasz refresh token, w odpowiedzi otrzymujesz nową parę:  access token (ważny kolejne 12h ),  refresh token (ważny kolejne 3 miesiące).  Czyli mam rozumieć że gdy uzyskam refresh token to tak naprawdę to znaczy że przez 3 miesiące mogę wywoływać pobranie nowego tokena który z kolei znów przed upływem 12 h muszę odświeżyć... jednym słowem nie oznacza to że zastosowanie refresh tokena sprawi że nic nie robię z nim przez 3 miesiące i zawsze działa bo mimo wszystko trzeba przeladowywac w ciągu tych 3 miesięcy grandem = refresh token   Tak ?  A po 3 miesiącach autoryzacja allegro wymagana o Dzięki temu nie musisz za każdym razem od początku się autoryzować. Czyli refresh token pozwala na uzyskanie nowego tokena dostępowego dla danego użytkownika bez konieczności ingerencji z jego strony.  Nie wiem czy można to jakoś zdebugować po waszej stronie ? jakiś trace-id ?  W tym konkretnym przypadku nie będzie takiej potrzeby. Oczywiście jeśli nadal masz wątpliwości co do całego procesu - jestem do dyspozycji :)  —  Reply to this email directly,  github.com view it on GitHub , or  github.com unsubscribe .  You are receiving this because you were mentioned. Message ID:  &lt;allegro/allegro-api/issues/6521/1368936042 @ github . com&gt;

**Comment by Lukasz-Zurek** [comment]
Tak, jeśli jednak generujesz nową parę tokenów, to czas 3 miesięcy żywotności (nowego) refresh_tokena, biegnie od początku. Jeśli natomiast w ciągu 3 miesięcy nie nastąpi użycie refresh_tokena - wymagana będzie nowa autoryzacja.

Z perspektywy twórcy aplikacji proces przedłużania ważności tokena jest tu bardzo istotny i jednocześnie przejrzysty z uwagi na **brak konieczności ingerencji ze strony użytkownika - nie musi on za każdym razem autoryzować się od początku.** To jest właśnie ta znacząca różnica.

**Comment by ircoo2022** [comment]
Ok czyli czy dobrze rozumiem że należy w cron co 12h uruchamiać przedłużanie tokena  otrzymują refresh token?     
            
          
     
      
       
        Dnia 3 stycznia 2023 07:30 Lukasz-Zurek ***@***.***&gt; napisał(a):
       
    
       
         
 Tak, jeśli jednak generujesz nową parę tokenów, to czas 3 miesięcy żywotności (nowego) refresh_tokena, biegnie od początku. Jeśli natomiast w ciągu 3 miesięcy nie nastąpi użycie refresh_tokena - wymagana będzie nowa autoryzacja. 
 Z perspektywy twórcy aplikacji proces przedłużania ważności tokena jest tu bardzo istotny i jednocześnie przejrzysty z uwagi na  brak konieczności ingerencji ze strony użytkownika - nie musi on za każdym razem autoryzować się od początku.  To jest właśnie ta znacząca różnica.

**Comment by Lukasz-Zurek** [comment]
@ircoo2022 Tak. Optymalnie najlepiej uruchamiać to krótko przed upływem 12h.

**Comment by ircoo2022** [comment]
ok i podajemy poprednio wygenerowany  token ( jako para z refresh_token ) jako parametr i otrzymujemy refresh_token ktorym sie poslugujemy dalej

**Comment by Lukasz-Zurek** [comment]
Chcąc wygenerować nową parę tokenów w ten sposób, przekazujesz tylko dotychczasowy refresh_token, tak jak to opisaliśmy [tutaj](https://developer.allegro.pl/tutorials/uwierzytelnianie-i-autoryzacja-zlq9e75GdIR#przedluzenie-waznosci-tokena).

**Comment by ircoo2022** [comment]
Ok dobrze  Czyli przekazuję dotychczasowy refreshToken  i posługuje się odebranym nowym refreshToken  
            
          
     
      
       
        Dnia 3 stycznia 2023 08:49 Lukasz-Zurek ***@***.***&gt; napisał(a):
       
    
       
         
 Chcąc wygenerować nową parę tokenów w ten sposób, przekazujesz tylko dotychczasowy refresh_token, tak jak to opisaliśmy  developer.allegro.pl tutaj .

**Comment by Lukasz-Zurek** [comment]
Tak. Posługujesz się nową odebraną parą tokenów. Za pomocą nowego access_token uzyskujesz dostęp do naszych zasobów, natomiast za pomocą nowego refresh_tokena generujesz nową parę tokenów.

**Comment by ircoo2022** [comment]
Mam pytanie czy podczas podpiania się pod produkt w Allegro ( produktyzacja ) - 'https://api.allegro.pl/sale/products?phrase='.$model.'&mode=GTIN', 

można coś takiego że wprowadzam kawałek nazwy i wyszukuje mi propozycje produktów allegro pod które mogę się podpiąć > wybieram któryś z listy. Czy można wybrać pod który produkt się podpinam czy tylko jeden zawsze ?

**Comment by Lukasz-Zurek** [comment]
@ircoo2022 Za pomocą wskazanego zasobu wyszukujesz produkty w naszym katalogu. Lista wyników wyszukiwania może być dłuższa, więc Ty decydujesz z którego identyfikatora produktu ostatecznie skorzystasz.

Natomiast jeśli chodzi o samo wystawianie ofert z produktem -zachęcam Cię mimo wszystko do korzystania z [POST /sale/product-offers](https://developer.allegro.pl/documentation/#operation/createProductOffers), gdzie proces przypisywania produktu do oferty jest częściowo zautomatyzowany. Więcej informacji znajdziesz w [naszym poradniku](https://developer.allegro.pl/tutorials/jak-jednym-requestem-wystawic-oferte-powiazana-z-produktem-D7Kj9gw4xFA#jak-wystawic-oferte-z-produktem-za-pomoca-zasobu-sale-product-offers).
